File: leetcode/Hot-100/c++/plt.cpp
#include <iostream>
class Solution {
// n超过45会超过int范围 溢出
public:
    int climbStairs1(int n) {
        int a = 1;
        int b = 1;
        int temp = 0;
        for (int i=0; i<n; i++) {
            temp = a + b;
            a = b;
            b = temp;
        }
        return a;
    }

    int climbStairs2(int n) {
        if (n < 3)
        {
            return n;
        }

        int f0 = 1;
        int f1 = 1;
        int f2 ;
        int i;
        for (i = 2; i <= n; i++)
        {
            f2 = f1 + f0;
            f0 = f1;
            f1 = f2;
        }
        return f2;
    }
};

int main() {
    Solution s;
    std::cout << s.climbStairs2(46) << std::endl;
}

File: leetcode/Hot-100/python/plt.py
"""
假设你正在爬楼梯。需要 n 阶你才能到达楼顶。
每次你可以爬 1 或 2 个台阶。你有多少种不同的方法可以爬到楼顶呢？
注意：给定 n 是一个正整数
例如：
    输入： 2
    输出： 2
    解释： 有两种方法可以爬到楼顶。
    1.  1 阶 + 1 阶
    2.  2 阶
"""
class Solution:
    def climbStairs(self, n: int) -> int:
        # 采用指针右移的方法 f(n) = f(n-1) + f(n-2)
        a = 1
        b = 1
        for i in range(n):
            temp = a + b
            a = b
            b = temp
        return a